### PR TITLE
CORS-2367: Gather bootstrap logs using proxy

### PR DESCRIPTION
**Pull request by EmilienM**
Implements: [CORS-2367](https://issues.redhat.com//browse/CORS-2367)

If SSH port is authorized in the list of rules to configure to which ports the proxy can establish a connection, it's possible to gather the bootstrap logs
through the proxy, for example in a disconnected environment.

```sh
openshift-install gather bootstrap --bootstrap ${BOOTSTRAP_HOST_IP} --proxy https://foo:mysecret@proxy.host
```

Co-Authored-By: Maysa De Macedo Souza <mdemaced@redhat.com>

### CODE REVIEW

File: pkg/gather/ssh/ssh.go [hunk 1]
```diff
@@ -13,10 +19,20 @@ import (
...
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 )
 
+// httpProxy is a HTTP/HTTPS connect proxy.
+type httpProxy struct {
+	host     string
+	haveAuth bool
+	username string
+	password string
+	forward  proxy.Dialer
```
Comment (MaysaMacedo) [inline pkg/gather/ssh/ssh.go:33]: can we use proxy.FromEnvironment? It seems to return a Dialer as well and might avoid the getHttpProxyUrl func

**Comment by EmilienM** [comment]
TODO:
* Make it work with the new patch made from Maysa's comments
* Test with `HTTPS_PROXY` (and a squid HTTPS endpoint)
* Add support for `NO_PROXY`
* Figure out if we need to add CLI args for proxy vars or should we read from env.

**Comment by EmilienM** [comment]
@patrickdillon could you look at this one please -- and tell us if we need to add CLI args for proxy vars or should we read from env, for UX.

@MaysaMacedo HTTPS doesn't work, we need to investigate.

**Comment by patrickdillon** [comment]
> @patrickdillon could you look at this one please -- and tell us if we need to add CLI args for proxy vars or should we read from env, for UX.

Typically we do not allow any env vars for user input. There are some exceptions to this, especially if we are using external packages that enable those env vars. In this case, I think env vars would be acceptable and sensible. 

As to env vars vs cli flags; I can't say I have a strong opinion there. My knee-jerk reaction is env vars, but that is pretty weakly held. Are you leaning one way or the other?

**Comment by EmilienM** [comment]
> As to env vars vs cli flags; I can't say I have a strong opinion there. My knee-jerk reaction is env vars, but that is pretty weakly held. Are you leaning one way or the other?

Both options have advantages and disadvantages:
* without CLI options, a user can just re-use exported environment
* exported environment can be forgotten when running CLI and not wanted at that time, so having CLI options makes it clear and wanted for the user that a proxy is needed to SSH the bootstrap.

No strong opinion here either.

**Comment by EmilienM** [comment]
@MaysaMacedo @patrickdillon I ended up adding a new argument for the CLI. This PR is ready for review now.

File: pkg/gather/ssh/ssh.go [hunk 2]
```diff
@@ -141,3 +191,84 @@ func LoadPrivateSSHKeys(paths []string) (map[string]interface{}, error) {
...
+// newHTTPSProxy creates a new httpProxy populated with credentials using HTTPS dialer
+func newHTTPSProxy(uri *url.URL, forward proxy.Dialer) (proxy.Dialer, error) {
+	return newHTTPProxy(uri, HttpsDialer)
+}
+
+// Dial is a custom dialer for the ssh client that will use HTTP CONNECT method via TLS handshake
+func (d httpsDialer) Dial(network, addr string) (c net.Conn, err error) {
+	c, err = tls.Dial("tcp", addr, TlsConfig)
+	if err != nil {
+		fmt.Println(err)
```
Comment (MaysaMacedo) [inline pkg/gather/ssh/ssh.go:219]: we probably want to wrap the error here with:

`return nil, errors.Wrap(err, "failed to connect to proxy")`

File: pkg/gather/ssh/ssh.go [hunk 3]
```diff
@@ -141,3 +191,84 @@ func LoadPrivateSSHKeys(paths []string) (map[string]interface{}, error) {
...
+	}
+	return
```
Comment (MaysaMacedo) [inline pkg/gather/ssh/ssh.go:221]: the default return should probably be:

`return c, nil`

File: pkg/gather/ssh/ssh.go (hunk 2 above)
Comment (MaysaMacedo) [reply pkg/gather/ssh/ssh.go:219]: Also, should we close the connection the same way is done with httpProxy?

File: pkg/gather/ssh/ssh.go [hunk 4]
```diff
@@ -141,3 +191,84 @@ func LoadPrivateSSHKeys(paths []string) (map[string]interface{}, error) {
...
+}
+
+// Dial is a custom dialer for the ssh client that will use HTTP CONNECT method
+// to establish a connection to the proxy. This is needed because the net/proxy module
+// only supports SOCKS5 proxies.
+func (s *httpProxy) Dial(network, addr string) (net.Conn, error) {
+	c, err := s.forward.Dial("tcp", s.host)
+	if c == nil {
```
Comment (MaysaMacedo) [inline pkg/gather/ssh/ssh.go:229]: is there a case when the connection is nil and no error was raised? I wonder if we can just check for the error

File: pkg/gather/ssh/ssh.go [hunk 5]
```diff
@@ -141,3 +191,84 @@ func LoadPrivateSSHKeys(paths []string) (map[string]interface{}, error) {
...
+	}
+
+	resp, err := http.ReadResponse(bufio.NewReader(c), req)
+	if err != nil {
+		c.Close()
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		logrus.Debugf("connect server using proxy, StatusCode [%d]", resp.StatusCode)
```
Comment (MaysaMacedo) [inline pkg/gather/ssh/ssh.go:267]: since we're returning the error, maybe there is no need for the log?

File: pkg/gather/ssh/ssh.go [hunk 6]
```diff
@@ -141,3 +191,84 @@ func LoadPrivateSSHKeys(paths []string) (map[string]interface{}, error) {
...
+}
+
+// newHTTPSProxy creates a new httpProxy populated with credentials using HTTPS dialer
+func newHTTPSProxy(uri *url.URL, forward proxy.Dialer) (proxy.Dialer, error) {
+	return newHTTPProxy(uri, HttpsDialer)
+}
+
+// Dial is a custom dialer for the ssh client that will use HTTP CONNECT method via TLS handshake
+func (d httpsDialer) Dial(network, addr string) (c net.Conn, err error) {
+	c, err = tls.Dial("tcp", addr, TlsConfig)
```
Comment (MaysaMacedo) [inline pkg/gather/ssh/ssh.go:217]: Is there a need for TlsConfig to contain RootCAs?

File: pkg/gather/ssh/ssh.go (hunk 1 above)
Comment (EmilienM) [reply pkg/gather/ssh/ssh.go:32]: So now we are providing the URL directly, so we can't use the Environment.
However I checked if we could use `proxy.FromURL` to return their Dialer, but this doesn't support HTTP Proxy (only Socks) - which is why we have to have the custom struct and function:
https://cs.opensource.google/go/x/net/+/refs/tags/v0.2.0:proxy/proxy.go;l=81-110

I'll close the comment but please re-open if you still think we can re-use something.

File: pkg/gather/ssh/ssh.go (hunk 6 above)
Comment (EmilienM) [reply pkg/gather/ssh/ssh.go:218]: Why would we need it? Do you refer to `additionalTrustBundle` somehow?

File: pkg/gather/ssh/ssh.go (hunk 2 above)
Comment (EmilienM) [reply pkg/gather/ssh/ssh.go:220]: done

File: pkg/gather/ssh/ssh.go (hunk 3 above)
Comment (EmilienM) [reply pkg/gather/ssh/ssh.go:222]: done

File: pkg/gather/ssh/ssh.go (hunk 5 above)
Comment (EmilienM) [reply pkg/gather/ssh/ssh.go:268]: right. i'll put the status code however.

File: pkg/gather/ssh/ssh.go (hunk 4 above)
Comment (EmilienM) [reply pkg/gather/ssh/ssh.go:230]: yes in fact if the proxy is unreachable (no route to host) we get no error.

**Comment by sadasu** [comment]
@EmilienM still interested in this?

**Comment by EmilienM** [comment]
I'll close it for now as none of us are actively working on that.
/close

**Pull request closed without merging**